[TG-3813] Allow specifying specific methods by regex to be loaded by lazy methods

thk123 opened the pull request:
Based on: #2335 :white_check_mark: 
Here we extend the existing option `--lazy-methods-extra-entry-point` to take a regex (previously allowed `.*` to be used to specify any method for a given class). This brings it into line with other "work around lazy loading not loading what I want". 

I maintained support for automatic descriptor filling (though now it will match all the methods whereas previously it would error). I also keep not requiring the `java::` prefix.

Commit:
CI lazy methods take vector of method loaders

These method loads can be used to manually load additional methods for
ci_lazy_methods

mgudemann commented on jbmc/regression/jbmc/lazyloading11/test.desc:
```diff
@@ -1,6 +1,6 @@
 CORE symex-driven-lazy-loading-expected-failure
 test.class
---lazy-methods --verbosity 10 --function test.f --lazy-methods-extra-entry-point 'test.sety:(I)V' --lazy-methods-extra-entry-point 'test.sety:(F)V'
+--lazy-methods --verbosity 10 --function test.f --lazy-methods-extra-entry-point 'test.sety:\(I\)V' --lazy-methods-extra-entry-point 'test.sety:\(F\)V'
```
are you sure this is needed in the quotes ?

thk123 replied:
You mean the escaping the brackets? Yes as otherwise it will be used as a regex group rather than matching the brackets (this is essentially what is changing).

mgudemann commented on jbmc/src/java_bytecode/ci_lazy_methods_needed.cpp:
```diff
@@ -43,3 +44,114 @@ bool ci_lazy_methods_neededt::add_needed_class(
 }
+
+/// Add to the needed classes all class specified, the replacement type if it
+/// will be replaced, and all feilds it contains.
```
... `fields` ...

thk123 replied:
Though this is a comment for #2335 so will squash into there

mgudemann commented on jbmc/src/java_bytecode/load_method_by_regex.cpp:
```diff
@@ -0,0 +1,70 @@
+
+  static const std::string java_prefix = "java::";
+  return descriptor_index == java_prefix.length() - 1 &&
+         has_prefix(pattern, "java::");
```
what does this check? the prefix check is done above, right?

thk123 replied:
Yes I worried about that being confusing but couldn't think of a better way - this is checking that if a `:` is found, it is _not_ the second colon in `java::`

thk123 commented:
CI failing for hopefully the reason in #2335 only, needs some unit tests for `load_method_by_regext`

LAJW commented on jbmc/src/java_bytecode/load_extra_methods.h:
```diff
@@ -0,0 +1,38 @@
+  {
+  }
+  std::vector<irep_idt>
+  extra_methods(const symbol_tablet &symbol_table) const override;
```
This class hierarchy appears to be unnecessary. You could greatly simplify all of that by using lambdas (and maybe higher-order functions to make your life easier).

LAJW commented on jbmc/src/java_bytecode/load_method_by_regex.cpp:
```diff
@@ -0,0 +1,70 @@
+    if(symbol.second.is_function())
+    {
+      const std::string symbol_name = id2string(symbol.first);
+      if(std::regex_match(symbol_name, regex))
```
Combine conditions, flatten the loop? Also this is another single-method-class that could've been replaced by a lambda.

thk123 replied:
Think is is clearer as two separate checks.

LAJW reviewed (commented):
I feel like the class hierarchy in the last two commits could be replaced by lambdas and functions returning lambdas. It would greatly reduce boilerplate.

Commit:
Add utility for checking two vectors are equal without caring about order

thk123 commented:
@LAJW repalaced class structure with lambda https://github.com/diffblue/cbmc/pull/2350/commits/cac9c7b1e988bc88b8a76f0b0808d05f98cad939

thk123 commented on jbmc/src/java_bytecode/load_method_by_regex.cpp:
```diff
@@ -0,0 +1,77 @@
+
```
This needs adding to a Makefile

thk123 commented:
TG pointer bump: https://github.com/diffblue/test-gen/pull/1960

mgudemann reviewed (approved):
LGTM

LAJW commented on jbmc/unit/java_bytecode/load_method_by_regex.cpp:
```diff
@@ -28,8 +19,7 @@ SCENARIO(
     {
-      bool result =
-        load_method_by_regex_testt::does_pattern_miss_descriptor(pattern);
+      bool result = does_pattern_miss_descriptor(pattern);
```
Nit `const` here and in the rest of this file.

LAJW commented on jbmc/src/java_bytecode/java_bytecode_language.cpp:
```diff
@@ -100,7 +100,7 @@ void java_bytecode_languaget::get_language_options(const cmdlinet &cmd)
     std::back_inserter(extra_methods),
     [](const std::string &pattern) {
-      return util_make_unique<load_method_by_regext>(pattern);
+      return build_load_method_by_regex(pattern);
```
Could you put `build_load_method_by_regex` as an argument of `std::transform` directly?

LAJW commented on jbmc/src/java_bytecode/load_method_by_regex.cpp:
```diff
@@ -28,43 +27,50 @@ load_method_by_regext::load_method_by_regext(const std::string &pattern)
+        if(std::regex_match(symbol_name, regex))
+        {
+          matched_methods.push_back(symbol_name);
+        }
```
The body of this loop could be simplified to:
```c++
if(symbol.second.is_function() && std::regex_match(id2string(symbol.first), regex))
   matched_methods.push_back(symbol.first);
```

thk123 replied:
Sure if you like, not convinced it is any more readable (And I'm not removing braces :upside_down_face: )

LAJW commented on unit/testing-utils/require_vectors_equal_unordered.h:
```diff
@@ -0,0 +1,28 @@
+template <class T>
+static void require_vectors_equal_unordered(
+  const std::vector<T> actual,
+  const std::vector<T> expected)
```
Remove `static` - it's an instruction for the linker that a function has been defined only in this file and nowhere else. While it might work on the first use (used only in one file at the moment), it might potentially break later when there are unresovled externals to an already defined specialization of this template.

Also, could you add `&` to `actual` and `expected`? As it is right now, this function will only work with copyable elements or rvalues. That means if you were testing a vector of unique pointers, it would only work with `std::move`.

LAJW reviewed (approved):
Some nitpicks about testing utilities. Please squash before merging, so that all these classes disappear.

Commit:
Extend lazy-methods-extra-entry-point to support regex methods

This will allow for loading for example, all constructors or all methods
called getSomething.

Tests using this require minor modifications when they have specified
the descriptor as the brackets must be escaped for the regex. However,
the new code handles the case where no descriptor is provided (in which
case it will match all overloads) and the case where no java:: prefix
has been provided, leaving similar behaviour.

Correcting documentation for the extra entry points

allredj reviewed (commented):
Passed Diffblue compatibility checks (cbmc commit: 783fceaf9f5ee1f60ff48ca8e003b5aadb811cb9).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/76972265
